library/bignum_mod_raw.c

```diff
@@ -119,7 +119,16 @@ int mbedtls_mpi_mod_raw_write( const mbedtls_mpi_uint *A,
 /* END MERGE SLOT 4 */
 
 /* BEGIN MERGE SLOT 5 */
-
+void mbedtls_mpi_mod_raw_add( mbedtls_mpi_uint *X,
+                              const mbedtls_mpi_uint *A,
+                              const mbedtls_mpi_uint *B,
+                              const mbedtls_mpi_mod_modulus *N )
+{
+    mbedtls_mpi_uint carry, borrow;
+    carry  = mbedtls_mpi_core_add( X, A, B, N->limbs );
+    borrow = mbedtls_mpi_core_sub( X, X, N->p, N->limbs );
+    (void) mbedtls_mpi_core_add_if( X, N->p, N->limbs, (unsigned) ( carry ^ borrow ) );
```

Comment:
Why are we casting out the return of this method? Shouldn't we at least make certain that core operation did what is expected?

Reply:
The return value is the carry, which we don't need.

The logic here is pretty much the same as used in many other places with modulus: 

If A + B < N, the add gives carry = 0, the sub gives borrow = 1, and we want to undo the subtraction, so we add-if with condition non-zero

If A + B >= N but < R (the radix/boundary), the add gives carry = 0, the sub gives borrow = 0, and we've actually got the value we want, so we don't want to add-if, and that's fine, since we give it a condition of 0

If A + B >= R, the add will give carry = 1, the subtraction will give borrow = 1 and we've got the value we want, so we don't want to undo it with the add-if, and that's fine since we give it a condition of 1 ^ 1, i.e. 0.